File: packages/client_sdk/typescript/src/realtime/wsClient.ts
/**
 * @file Gateway realtime WebSocket client.
 *
 * Subscribes to live control-plane events (invalidate, tasks_update,
 * connected, …). Reconnect and handler registration live here so React
 * hooks stay declarative.
 */

import type { GatewayRealtimeMessage } from './events'

/** interface `GatewayRealtimeClientOptions` — Gateway client SDK contract type. */
export interface GatewayRealtimeClientOptions {
  /** Example: "ws://localhost:8000" (no trailing slash). Empty string means relative WS URL. */
  wsBaseUrl?: string
}

/** class `GatewayRealtimeClient` — Gateway client SDK helper; see methods. */
export class GatewayRealtimeClient {
  private wsBaseUrl: string

  constructor(opts?: GatewayRealtimeClientOptions) {
    this.wsBaseUrl = opts?.wsBaseUrl ?? ''
  }

  connectGlobalUpdates(onMessage: (msg: GatewayRealtimeMessage) => void): WebSocket {
    /** `url` — Gateway client SDK export. */
    const url = `${this.wsBaseUrl}/ws/global-updates`
    /** `ws` — Gateway client SDK export. */
    const ws = new WebSocket(url)
    ws.onmessage = (ev) => {
      try {
        onMessage(JSON.parse(ev.data))
      } catch {
        // ignore
      }
    }
    // Return this value to the Gateway client caller.
    return ws
  }

  connectPlanExecution(planId: number, onMessage: (msg: GatewayRealtimeMessage) => void): WebSocket {
    /** `url` — Gateway client SDK export. */
    const url = `${this.wsBaseUrl}/ws/execution/${planId}`
    /** `ws` — Gateway client SDK export. */
    const ws = new WebSocket(url)
    ws.onmessage = (ev) => {
      try {
        onMessage(JSON.parse(ev.data))
      } catch {
        // ignore
      }
    }
    // Return this value to the Gateway client caller.
    return ws
  }
}

File: packages/client_sdk/python/src/realtime/ws_client.py
from __future__ import annotations

"""Module ``client_sdk/python/src/realtime/ws_client.py``.

WebSocket / realtime event clients for Gateway live updates.

Part of the Gateway client SDK: callers use HTTP and WebSocket helpers to talk to the BFF without coupling to fleet_server gRPC.

Hand-written source for ``ws_client.py``. Behavior is unchanged; comments document control-plane / agent-runtime intent for maintainers.
"""


import json
from typing import Callable, Optional

import websockets

from .events import GatewayRealtimeMessage


class GatewayRealtimeClient:
    """``GatewayRealtimeClient`` — agent_fleet packages helper; see body comments for step-by-step behavior."""
    def __init__(self, ws_base_url: str | None = None):
        """``GatewayRealtimeClient`` — agent_fleet packages helper; see body comments for step-by-step behavior."""
        if not ws_base_url:
            from packages.platform_config import setting
            ws_base_url = setting("GATEWAY_WS_URL")
        self.ws_base_url = ws_base_url.rstrip("/")

    async def watch_global_updates(self, on_message: Callable[[GatewayRealtimeMessage], None]) -> None:
        """``__init__`` — agent_fleet packages helper; see body comments for step-by-step behavior."""
        url = f"{self.ws_base_url}/ws/global-updates"
        # Hold ``websockets.connect(url)`` for the duration of the indented block.
        async with websockets.connect(url) as ws:
            # Loop: async for msg in ws.
            async for msg in ws:
                # Try the fallible work below.
                try:
                    # Call ``on_message``.
                    on_message(json.loads(msg))
                # On except Exception: recover or re-raise as appropriate.
                except Exception:
                    pass

    async def watch_plan_execution(self, plan_id: int, on_message: Callable[[GatewayRealtimeMessage], None]) -> None:
        """``watch_plan_execution`` — agent_fleet packages helper; see body comments for step-by-step behavior."""
        url = f"{self.ws_base_url}/ws/execution/{plan_id}"
        # Hold ``websockets.connect(url)`` for the duration of the indented block.
        async with websockets.connect(url) as ws:
            # Loop: async for msg in ws.
            async for msg in ws:
                # Try the fallible work below.
                try:
                    # Call ``on_message``.
                    on_message(json.loads(msg))
                # On except Exception: recover or re-raise as appropriate.
                except Exception:
                    pass

